src/java.base/share/classes/com/sun/crypto/provider/CipherCore.java

```diff
@@ -748,12 +748,26 @@ byte[] doFinal(byte[] input, int inputOffset, int inputLen)
         throws IllegalBlockSizeException, BadPaddingException {
         try {
             byte[] output = new byte[getOutputSizeByOperation(inputLen, true)];
-            byte[] finalBuf = prepareInputBuffer(input, inputOffset,
-                    inputLen, output, 0);
+            int outputOffset = 0;
+            int outLen = 0;
+
+            if (!decrypting && (padding != null) && (inputLen > 0)) {
```

Comment:
Why only when encrypting?

Reply:
Because when decrypting is performed, or padding is not required, the input is not copied into an intermediate buffer. In that case, there is no point in calling `update()` and doing an extra trip to the native code with all the issues and delays that might be causing, since there is nothing to be gained.

Reply:
That's not the way I interpret the condition in `prepareInputBuffer()` following the comment:
```java
        /*
         * prepare the final input, assemble a new buffer if any
         * of the following is true:
         *  - 'input' and 'output' are the same buffer
         *  - there are internally buffered bytes
         *  - doing encryption and padding is needed
         */
```